### DISCUSSION TITLE
Proposal: Port to Rust

### DISCUSSION THREAD

**Discussion by fosskers** [fosskers/aura#657] [Ideas]
With Aura nearing its 10th birthday, I'd like to open a discussion about the possibility of porting Aura to the [Rust] language. This has been on my mind for some time, but I don't believe that I'm being swept up in hype. Nor do I intend this discussion to be one of "Haskell vs Rust" in general. Both are fine languages. 

Such a change would represent Aura 4.0.0, and let me be clear that as far as users are concerned, they shouldn't notice any difference in Aura's commands / data formats / config formats. However when it comes to authorship and maintenance of Aura (which are primarily my personal burden), Rust as a systems language offers some distinct advantages.

### PROs

- **Access to `libalpm`:** This would improve performance in a number of places
  and unlock new features that were previously unfeasable. This is the form that
  I always wanted Aura to have, instead of relying on `pacman` so much.
- **More opportunity for outside contributions:** There are simply more Rust
  devs than Haskell devs.
- Better performance, smaller binaries, and ~static linking~ (we'll likely have to stay dynamic, due to `libalpm`).
- Official ecosystem support for release candidates. Overall easier to distribute.
- The ecosystem has much more momentum and will likely suffer less bitrot.
- Builds on more platforms/architectures (re: Macbooks will soon be using ARM
  chips. Related [stack issue]).
- Access to [clap] for CLI args:
  - **Can autogenerate completions.**
  - [Supports "fused short ops"](https://github.com/clap-rs/clap/blob/master/examples/23_flag_subcommands_pacman.rs) (that Pacman and Aura have) out of the box.
  - Offers suggestions if users make a typo.
  - Can define argument groups for the subflags of the big commands (`-S`, `-Q`,
    etc.). **This should fix the currently garbled `--help` output for free.**
- Error handling would be more straight-forward (re: IO exceptions in Haskell).
- Saner `String` handling.
- #572 #625 #640 would be fixed for free. 
- A rewrite acts as a "spring cleaning" of sorts, and would clear away historical cruft.
- Can get rid of the excess config.
- Aura would be versioned via SemVer instead of PVP.

### CONs

- I risk upsetting long-time Haskell users.
- Potential conflict with the [rua] project, although Aura is a drop-in
  replacement for Pacman, where `rua` is just for building AUR packages.
- Effort of porting the [language strings](https://github.com/fosskers/aura/blob/073740f60bfc1b3911c5312c2d96d624b8de84f5/aura/lib/Aura/Languages.hs).
- Must write a bash parser to replace the `language-bash` Haskell library.
  However this should be a straight-forward port, and I can probably claim the
  [bash] crate.

So, I ask for the input of concerned community members. Is there anything I'm not accounting for? I'm willing to the put the time into such a port to Rust, and think that doing so would be worth it in the long-term.

[Rust]: https://www.rust-lang.org/
[clap]: https://lib.rs/crates/clap
[rua]:  https://lib.rs/crates/rua
[bash]: https://lib.rs/crates/bash
[stack issue]: https://github.com/commercialhaskell/stack/issues/2103

**Comment by ratijas** [comment]
_Assuming today is not a 1th of April…_

Whoa... That's one hell of an ambition.

> Access to clap for CLI args

Actually, I've seen counter-examples when people [got rid](https://blog.kodewerx.org/2020/06/the-rust-compiler-isnt-slow-we-are.html) of bloated `clap`, to get their number of dependencies and binary size cut down in half.

> Must write a bash parser to replace the language-bash Haskell library.
> However this should be a straight-forward port, and I can probably claim the
> bash crate.

I briefly looked at it, and I failed to understand what it is or why it is needed. Are there external commands invoked from aura where `exec(["vec", "of", "args"])` is not enough?

If anything, this old but good book might come in handy:

![Rewriting it in Rust](https://user-images.githubusercontent.com/6737986/97060700-da828380-159c-11eb-8e52-1e087cae48c0.png)

**Comment by fosskers** [comment]
`clap` is definitely a chubby dep, and in all my other Rust projects I use [gumdrop](https://lib.rs/crates/gumdrop). `clap`'s full-featuredness is very appealing, in this case. And the fact that it can do the "fused short-ops" like `-Syu` out-of-the-box means I wouldn't have to write extra code. 

As for Bash, some other AUR-related tools use `shellcheck` to scan for PKGBUILD safety, but this isn't enough in my opinion. PKGBUILD exploits take very specific forms, and Aura knows how to check for these. Parsing the bash (without running it) is the first step of that process.

And when it comes to the effort of the port itself - luckily Aura is already built and has a decade of accumulated lessons :) I don't need to reinvent the logic itself, I just need to transfer it. I already know Rust, so there's no extra overhead there either.

**Comment by fosskers** [comment]
One feature that I've wanted for a long time, which `libalpm` would unlock: Hijacking `-Su` to do concurrent downloads.

**Comment by ratijas** [comment]
> One feature that I've wanted for a long time, which `libalpm` would unlock: Hijacking `-Su` to do concurrent downloads.

That would be very good. Although, as far as I can remember, I explored pacman source code, and I saw there a simple loop written in plain C, which iterates over packages and downloads them one by one. But if that loop can be separated from the download routine without exclusively locking the database, that would be so cool.

**Comment by fosskers** [comment]
Yes I've wanted that for years. It's hard in C, but trivial in a modern language like Rust. [indicatif](https://lib.rs/crates/indicatif) can be used to show multithreaded progress, as well!

**Comment by ratijas** [comment]
> indicatif can be used to show multithreaded progress, as well!

Yes, it would require more elaborate progress indicator than simple `"\r[" + "=" * width"`, more like something in `docker pull`. Examples on `indicatif` page are quite shiny, but I'd rather stick to plain `[==> ]` style to keep it as close to pacman as possible. Fortunately, it seems to support pretty elaborate styling.

**Comment by fosskers** [comment]
I've used it before and it works quite well. We'd have full control over the styling, so `ILoveCandy` should still be possible :)

**Comment by ratijas** [comment]
> `ILoveCandy`

God save the Queen... :open_mouth:

**Comment by fosskers** [comment]
Update:
- `clap` seems to do everything I need, but I found [an issue with fused short opts](https://github.com/clap-rs/clap/discussions/2185) that may or may not be a bug.
- I'm able to link to `libalpm` and interact with it!
- Unfortunately, I can't seem to static link `libalpm` with my usual technique of invoking: `--target x86_64-unknown-linux-musl`

**Comment by fosskers** [comment]
![concurrent-downloads](https://user-images.githubusercontent.com/229679/97510530-24bf8800-1942-11eb-8d85-d06d8df5ef52.gif)

This is a demonstration of my Rust port proof-of-concept. It:
- makes multiple connections to `libalpm` via `r2d2` and queries it for package information
- downloads the corresponding package tarballs via `curl` and `rayon`
- shows concurrent progress via `indicatif`

**Comment by simonsan** [comment]
Sounds like a really good idea to me! 👍

**Comment by fosskers** [comment]
The CLI flag definitions have been vastly simplified due to auto-derivation of the `Clap` trait:
```
> cargo run -- -h
    Finished dev [unoptimized + debuginfo] target(s) in 0.05s
     Running `target/debug/aura -h`
aura 4.0.0-r1
Colin Woodbury
Install and manage Arch Linux and AUR packages

USAGE:
    aura [FLAGS] <SUBCOMMAND>

FLAGS:
        --english     Output in English
    -h, --help        Prints help information
        --japanese    Output in Japanese
    -q, --quiet       Less verbose output
    -V, --version     Prints version information

SUBCOMMANDS:
    -G, completions     Generate shell completions for Aura
    -S, --sync, sync    Synchronize packages
```
Language flags conflict:
```
> cargo run -- -Si linux --english --japanese
    Finished dev [unoptimized + debuginfo] target(s) in 0.04s
     Running `target/debug/aura -Si linux --english --japanese`
error: The argument '--english' cannot be used with '--japanese'

USAGE:
    aura {sync, --sync, -S} [FLAGS] [packages]...

For more information try --help
```
You'll also notice `-G`, a demonstration of auto-generation of Bash/Zsh/Fish completions. No more maintaining them by hand!

Further, sane and organized localized output is now possible via [fluent](https://lib.rs/crates/fluent), the Rust client for [Project Fluent](https://www.projectfluent.org/). No more baking the translations into the binary!

**Comment by fosskers** [comment]
Another advantage of Rust is the ability to upload release candidates. I could provide a third package on the AUR, `aura-beta`, that would offer access to these.

**Comment by fosskers** [comment]
I have a solution for Aura's debug logging, configurable via the command line as expected:
```
> cargo run
    Finished dev [unoptimized + debuginfo] target(s) in 0.04s
     Running `target/debug/aura`
aura 4.0.0-r1
Colin Woodbury
Install and manage Arch Linux and AUR packages

USAGE:
    aura [FLAGS] [OPTIONS] <SUBCOMMAND>

FLAGS:
        --english     Output in English
    -h, --help        Prints help information
        --japanese    Output in Japanese
    -q, --quiet       Less verbose output
    -V, --version     Prints version information

OPTIONS:
        --log-level <LEVEL>    Minimum level of log messages to display [possible values: debug,
                               info, warn, error]

SUBCOMMANDS:
    -G, completions     Generate shell completions for Aura
    -S, --sync, sync    Synchronize packages
```
The neat part is that the logger implementation doesn't need to leak into the library, meaning different UIs could choose different backends.

Stripped release-mode binary size is 1.7mb at this point. Adding `codegen-units = 1` doesn't really affect the result, but adding `opt-level = 'z'` drops it to 1.4mb.

`pacman` itself is 144k, while `yay` is 7.9mb, and the current Haskell implementation of Aura is 8.9mb.

**Comment by Morganamilo** [comment]
I'd have to recommend against clap if you're still going to be a pacman wrapper. Clap likes to do things a certain way, and trying to force it to behave exactly like pacman is a big pain,

**Comment by Morganamilo** [comment]
> One feature that I've wanted for a long time, which `libalpm` would unlock: Hijacking `-Su` to do concurrent downloads.

This is already in pacman-git. Will hit in the next release.

**Comment by Morganamilo** [comment]
> Update:
> 
> * `clap` seems to do everything I need, but I found [an issue with fused short opts](https://github.com/clap-rs/clap/discussions/2185) that may or may not be a bug.
> * I'm able to link to `libalpm` and interact with it!
> * Unfortunately, I can't seem to static link `libalpm` with my usual technique of invoking: `--target x86_64-unknown-linux-musl`

Arch's pacman does not provide a static libalpm.a to link against. Although you can compile it yourself to get one. And then hack alpm.rs to support it.

**Comment by Morganamilo** [comment]
> However this should be a straight-forward port, and I can probably claim the
bash crate.

Yeah don't bet on it. That guy has claimed 100s of common names and doesn't give them out despite the desc. I wanted the crate yay ;-;

**Comment by fosskers** [comment]
Thanks for jumping into the conversation :)

> I'd have to recommend against clap if you're still going to be a pacman wrapper. Clap likes to do things a certain way, and trying to force it to behave exactly like pacman is a big pain,

I believe I've already succeeded :smile: 

> This is already in pacman-git. Will hit in the next release.

Hot damn!

> Arch's pacman does not provide a static libalpm.a to link against. Although you can compile it yourself to get one. And then hack alpm.rs to support it.

Yeah looks like I'll be sticking with dynamic linking for the near future.

> Yeah don't bet on it. 

He's squatting on some real nice names eh? Okay, plan B then.

**Comment by simonsan** [comment]
> He's squatting on some real nice names eh? Okay, plan B then.

Just send an eMail to help@crates.io and try to claim the crate you want. I think how this guy is doing it, was not the intention. Worth a try. ;-) I already sent an eMail to bring some attention to this special case of name squatting.

**Comment by Morganamilo** [comment]
> I believe I've already succeeded

I'd love to get my hands on the code and try break it :p

**Comment by fosskers** [comment]
> I'd love to get my hands on the code and try break it :p

I'll have the new PR up soon.

**Comment by mb720** [comment]
This is exciting! 👍
>Access to libalpm

I'm curious, why wasn't it possible to access `libalpm` using Haskell?

**Reply by fosskers** [reply]
No one has written bindings for it, so more than once I tried to myself. This was a few years ago, and I could never quite find a good dev workflow with the existing tooling. Rust on the other hand has [fantastic bindings](https://lib.rs/crates/alpm) thanks to @Morganamilo , and they're done in such a way as to be zero-cost (i.e. very efficient).  

Although at this point, even if `alpm` access were possible through Haskell, I'd still switch.

**Comment by Morganamilo** [comment]
Almost zero cost. Sadly converting from rust strings to c strings requires adding a null byte to the end. But I've done my best to have as little overhead as possible.

**Comment by poscat0x04** [comment]
Just FYI I have written an experimental Haskell [binding](https://github.com/poscat0x04/alpm4h) to libalpm :)

**Comment by micwoj92** [comment]
So after switching to rust will there only be this rust port or the haskell version will still be supported for bug fixes only?

**Reply by fosskers** [reply]
Yes, my plan is to release an `aura3-bin` package on the AUR that offers Haskell-Aura as it currently is. Bug fixes only.

**Comment by unhammer** [comment]
@fosskers Out of curiosity, why were completions maintained by hand instead of generated by [optparse-applicative](https://hackage.haskell.org/package/optparse-applicative-0.16.1.0#bash-zsh-and-fish-completions)?

**Reply by fosskers** [reply]
Ah the last time I had checked that wasn't an out-of-the-box feature of `optparse-applicative`, but admittedly said checking was a few years ago.

**Comment by ShalokShalom** [comment]
Is there anything left, which is currently hindering an early release?

**Reply by fosskers** [reply]
I recently completed `-A` and a new config format. In my mind, what's left before an initial `4.0.0-alpha` could be released would be tighter detection of various pacman usecases (like `-Syu`), and giving them special treatment. I will be starting that work next week.

**Reply by ShalokShalom** [reply]
Huuray, I am waiting for this. :)

**Reply by ShalokShalom** [reply]
Still waiting :P

**Reply by fosskers** [reply]
I recently moved across the world and started a new job, but this is still high on my priority list.

**Reply by ShalokShalom** [reply]
Wonderful, congratulations 🥳

**Reply by fosskers** [reply]
I have just released `aura-git` to the AUR which targets the `4.0.0` RC1: https://aur.archlinux.org/packages/aura-git

**Reply by micwoj92** [reply]
Is `fd` and `ripgrep` really hard dependency? I don't use these tools and would like to not have to install this. Could fallback be made to "regular" stuff when these are not present?

**Reply by fosskers** [reply]
Thanks for the feedback, I should be able to do that replacement.

**Reply by fosskers** [reply]
Done.

**Comment by ShalokShalom** [comment]
Is this complete now? As I dont see a release yet.

**Reply by fosskers** [reply]
We're at `alpha8` currently. I'm finishing up some work with the [Faur](https://git.sr.ht/~fosskers/faur) server at the moment, and once that's done I plan to return to Aura development proper.

**Comment by ShalokShalom** [comment]
Considering there had been no release since 2 years, and you planned to release an early Rust version for just as long: Is there anything major, that causes such huge delays?

**Reply by fosskers** [reply]
Actually you're free to install the Beta at any time:
```
cargo install aura-pm --version 4.0.0-beta02
```
Delays are otherwise caused by life. I plan to do an official release in June.

**Reply by ShalokShalom** [reply]
wonderful
